### DISCUSSION TITLE
Why setting label to 0 for all data in gen_data_set_youtube()?

### DISCUSSION THREAD

**Discussion by Wp-Zhang** [bbruceyuan/DeepMatch-Torch#1] [Q&A] [answered]
I'm new to this area so it may be a dumb question.
When generating the dataset for YoutubeDNN [here](https://github.com/bbruceyuan/DeepMatch-Torch/blob/main/examples/preprocess.py), why do we need to set the label for the whole dataset as 0 as the code below instead of setting 1 to positive items and 0 to sampled negative items which is the method in the Tensorflow version of DeepMatch?
` train_set.append((reviewerID, hist[::-1], [pos_list[i]] + [neg_list[item_idx] for item_idx in np.random.choice(neg_list, negsample)], 0, len(hist[::-1]),rating_list[i]))`

**Comment by bbruceyuan** [comment, marked as answer]
hi, it is a good question and I think about this question a long time too. 

English Version:
As we all know, there is no sampled_softmax_loss like tf. Then I use CrossEntropyLoss instead of sampled_softmax_loss through constructing sample label [1, 0, 0, 0, 0], and set  postivate sample displayed in first position, all 0 is sampled negative sample. 

Additionally, in [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html), `Target` shape is (N), so label should be  0 (the index of this positive sample).

my code can see: PLBaseModel.py `_step` function. https://github.com/bbruceyuan/DeepMatch-Torch/blob/322589cbe7f6d4efb1c4094bd6fb62f599f62c29/deepmatch_torch/models/PLBaseModel.py


If you have a better pytorch version `sampled_softmax_loss` implementation，can you seed me a pr to construct it better?

Chinese Version:
在 PyTorch 里面，没有 sampledsoftmax_loss 函数，所以我通过 自己构造样本 [1, 0, 0, 0, 0] 进行 CrossEntropyLoss 来代替 sampled_softmax_loss, 1是正样本， 0 是采样的, 让 第一个元素用于是 正样本，这样进行交叉熵可以起到类似 sampledsoftmax 的作用。

那么在 PyTorch 里面的 [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) 的输入参数是 Target, shape 是 (N)；因为我们用交叉熵做 loss, 所以 label 是第一个位置，因此 label 是 0。


代码部分可以查看：https://github.com/bbruceyuan/DeepMatch-Torch/blob/322589cbe7f6d4efb1c4094bd6fb62f599f62c29/deepmatch_torch/models/PLBaseModel.py 中 _step 函数；
